### PR TITLE
Make importer save an uncompressed form for performance.

### PR DESCRIPTION
**Pull request by mattmoor**
This re-exports our internal copy of: https://github.com/google/containerregistry/pull/28

### CODE REVIEW

**Comment by mattmoor** [comment]
I think the main change is moving the new methods onto the `DockerImage` interface so that they can be used regardless of form.

I opted to drop the flag because more or less the sole purpose of the importer is `container_load`, which (when we switch `rules_docker` to contain this) will exclusively opt into the uncompressed form.

I also renamed the new layer accessor to `[uncompressed_]layer`.  It's not perfect, but hopefully the doc comments demystify things...  I'm definitely open to suggestions.

**Comment by mattmoor** [comment]
@jonjohnsonjr @damienmg @dlorenc FYI

File: client/v2_2/docker_image_.py [hunk 1]
```diff
@@ -105,6 +110,27 @@ def uncompressed_blob(self, digest):
...
     return unzipped
 
+  def _diff_id_to_digest(self, diff_id):
+    for (this_digest, this_diff_id) in zip(self.fs_layers(), self.diff_ids()):
+      if this_diff_id == diff_id:
+        return this_digest
+    raise ValueError('Unmatched "diff_id": "%s"' % diff_id)
+
+  def layer(self, diff_id):
+    """Like `blob()`, but accepts the uncompressed name `diff_id` instead.
```
Comment (siddharthab) [inline client/v2_2/docker_image_.py]: nit: "uncompressed name \`diff_id\`" is not very clear to me.

File: client/v2_2/save_.py [hunk 2]
```diff
@@ -185,3 +185,72 @@ def write_file(
...
+    for diff_id in reversed(image.diff_ids()):
+      # Create a local copy
+      digest_name = os.path.join(directory, '%03d.sha256' % idx)
+      f = executor.submit(write_file, digest_name,
+                          # Strip the sha256: prefix
+                          lambda diff_id: diff_id[7:], diff_id)
+      future_to_params[f] = digest_name
+
+      layer_name = os.path.join(directory, '%03d.tar' % idx)
+      f = executor.submit(write_file, layer_name, image.layer, diff_id)
```
Comment (siddharthab) [inline client/v2_2/save_.py]: This should be image.uncompressed_layer.

**Comment by siddharthab** [comment]
Sorry I did not notice that you had updated the branch.

The current code works like a charm. Thank you so much for your very helpful support with this request. From my point of view, we can merge this. 😄 

Some observations when I was testing with .tar.gz archives (not saying that we have to do anything about it, just FYI):
- The archive is uncompressed multiple times, once each for the manifest, the config file, and each layer. The time for each uncompression is almost the same as `tar.extractall` on the same archive.
- Single-threaded execution is faster than multi-threaded. 11s vs 14s for ubuntu:16.04 image on my machine.

All tests were done with
```
docker save ubuntu:16.04 | gzip > ubuntu.tar.gz
mkdir ubuntu
time bazel run :importer -- --directory ubuntu --tarball ubuntu.tar --stderrthreshold INFO
```

**Comment by mattmoor** [comment]
I'm going to bundle this into a full fast-forward from the mono-repo

**Pull request closed without merging**